Change program dependencies to cadence location

janezpodhostnik opened the pull request:
ref: https://github.com/onflow/cadence/issues/1684

This is now possible that account freezing is removed, and will be needed to remove double counting of imports.

pattyshack reviewed (changes requested):
I think this breaks "*" import (loading vs not loading an unrelated program in the same address will result in identical cadence execution, but different spock).  Can you add a test to ensure that's not the case?

janezpodhostnik commented:
Thank you for pointing that out. It was problematic.

I still need dependencies per location for the next step. So I changed to track dependencies by both address and location and split invalidation to handle contract deployment differently from contract additions and removals.

pattyshack commented on fvm/environment/derived_data_invalidator.go:
```diff
@@ -101,14 +111,34 @@ func (invalidator ProgramInvalidator) ShouldInvalidateEntry(
-	for _, key := range invalidator.ContractUpdateKeys {
-		_, ok := program.Dependencies[key.Address]
+	for _, key := range invalidator.ContractUpdates.Updates {
+		loc := common.AddressLocation{
```
recreating the address location object on each ShouldInvalidateEntry call is expensive.  maybe do some pre-processing.  similar below for common.Address

janezpodhostnik replied:
I will address this in a separate PR when I switch everything to Address Location https://github.com/onflow/cadence/pull/2382

pattyshack commented on fvm/environment/contract_updater.go:
```diff
@@ -437,26 +437,50 @@ func (updater *ContractUpdaterImpl) RemoveContract(
+			if err != nil {
+				return ContractUpdates{}, err
+			}
+		} else if currentlyExists {
```
maybe combine the else if / else.  ContractExists don't need to be called when deleting.  SetContract can be call unconditionally

janezpodhostnik commented:
bors merge

janezpodhostnik commented:
bors merge
